// ng-go-top-button/esm2015/lib/ng-go-top-button.module.js
import { NgGoTopButtonComponent } from './ng-go-top-button.component';
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import * as ɵngcc0 from '@angular/core';
export class GoTopButtonModule {
}
GoTopButtonModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: GoTopButtonModule });
GoTopButtonModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function GoTopButtonModule_Factory(t) { return new (t || GoTopButtonModule)(); }, imports: [[CommonModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(GoTopButtonModule, { declarations: function () { return [NgGoTopButtonComponent]; }, imports: function () { return [CommonModule]; }, exports: function () { return [NgGoTopButtonComponent]; } }); })();
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(GoTopButtonModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule],
                declarations: [NgGoTopButtonComponent],
                exports: [NgGoTopButtonComponent]
            }]
    }], null, null); })();
